nexus/db-queries/src/db/datastore/affinity.rs

```diff
@@ -0,0 +1,2701 @@
+use omicron_common::api::external::Error;
+use omicron_common::api::external::ListResultVec;
+use omicron_common::api::external::LookupType;
+use omicron_common::api::external::ResourceType;
+use omicron_common::api::external::UpdateResult;
+use omicron_uuid_kinds::AffinityGroupUuid;
+use omicron_uuid_kinds::AntiAffinityGroupUuid;
+use omicron_uuid_kinds::GenericUuid;
+use omicron_uuid_kinds::InstanceUuid;
+use ref_cast::RefCast;
+
+impl DataStore {
+    pub async fn affinity_group_list(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+        pagparams: &PaginatedBy<'_>,
+    ) -> ListResultVec<AffinityGroup> {
+        use db::schema::affinity_group::dsl;
+
+        opctx.authorize(authz::Action::ListChildren, authz_project).await?;
+
+        match pagparams {
+            PaginatedBy::Id(pagparams) => {
+                paginated(dsl::affinity_group, dsl::id, &pagparams)
+            }
+            PaginatedBy::Name(pagparams) => paginated(
+                dsl::affinity_group,
+                dsl::name,
+                &pagparams.map_name(|n| Name::ref_cast(n)),
+            ),
+        }
+        .filter(dsl::project_id.eq(authz_project.id()))
+        .filter(dsl::time_deleted.is_null())
+        .select(AffinityGroup::as_select())
+        .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    pub async fn anti_affinity_group_list(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+        pagparams: &PaginatedBy<'_>,
+    ) -> ListResultVec<AntiAffinityGroup> {
+        use db::schema::anti_affinity_group::dsl;
+
+        opctx.authorize(authz::Action::ListChildren, authz_project).await?;
+
+        match pagparams {
+            PaginatedBy::Id(pagparams) => {
+                paginated(dsl::anti_affinity_group, dsl::id, &pagparams)
+            }
+            PaginatedBy::Name(pagparams) => paginated(
+                dsl::anti_affinity_group,
+                dsl::name,
+                &pagparams.map_name(|n| Name::ref_cast(n)),
+            ),
+        }
+        .filter(dsl::project_id.eq(authz_project.id()))
+        .filter(dsl::time_deleted.is_null())
+        .select(AntiAffinityGroup::as_select())
+        .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    pub async fn affinity_group_create(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+        group: AffinityGroup,
+    ) -> CreateResult<AffinityGroup> {
+        use db::schema::affinity_group::dsl;
+
+        opctx.authorize(authz::Action::CreateChild, authz_project).await?;
+
+        let conn = self.pool_connection_authorized(opctx).await?;
+        let name = group.name().as_str().to_string();
```

Comment:
It seems like it would be nice if we just had a way to unwrap an owned `Name` into its inner `String` without having to do this dance of getting a reference to it and copying that `&str` into a new `String`. Not something we have to solve in this PR though...

Reply:
I can add a conversion function, but I think I still need to clone the `group` containing this name to pass it into `diesel::insert_into`